experimental/slo/slo_middleware.go

```diff
@@ -0,0 +1,47 @@
+package slo
+
+import (
+	"errors"
+	"net/http"
+	"syscall"
+	"time"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana-plugin-sdk-go/backend/httpclient"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+var duration = promauto.NewHistogramVec(prometheus.HistogramOpts{
+	Namespace: "grafana",
+	Name:      "plugin_external_requests_duration",
+	Help:      "Duration of requests to external services",
+}, []string{"plugin", "error_source"})
+
+// Middleware captures duration of requests to external services and the source of errors
+func Middleware(plugin string) httpclient.Middleware {
+	return httpclient.NamedMiddlewareFunc(plugin, func(opts httpclient.Options, next http.RoundTripper) http.RoundTripper {
+		return RoundTripper(plugin, opts, next)
+	})
+}
+
+// RoundTripper captures duration of requests to external services and the source of errors
+func RoundTripper(plugin string, _ httpclient.Options, next http.RoundTripper) http.RoundTripper {
+	return httpclient.RoundTripperFunc(func(req *http.Request) (*http.Response, error) {
+		start := time.Now()
+		var errorSource = "none"
+
+		defer func() {
+			duration.WithLabelValues(plugin, errorSource).Observe(time.Since(start).Seconds())
+		}()
+
+		res, err := next.RoundTrip(req)
+		if res != nil && res.StatusCode >= 400 {
+			errorSource = string(backend.ErrorSourceFromHTTPStatus(res.StatusCode))
+		}
+		if errors.Is(err, syscall.ECONNREFUSED) {
```

Comment:
Nit. Isn't any error a downstream error source?

Reply:
it depends really.  this is setting defaults.  but something like StatusNotImplemented would suggest there is something wrong in the plugin.  it's calling the wrong endpoint, or an endpoint changed.  so the plugin needs to be fixed